### PR TITLE
add slow path for is_contiguous

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77396

Implementation is done by roughly following the template from [Issue 65423](https://github.com/pytorch/pytorch/issues/65423)
- Adding an argument `uses_python_dispatch` (defaulted to False) to `_make_subclass` and `_make_wrapper_subclass`
- If True, then I'll `set_python_dispatch(true)` and set `sizes_strides_policy_` to `SizesStridesPolicy::CustomStrides`
- In `is_contiguous`, if both of the above are true, then I'll call the new method on PyInterpreter

This should avoid slowing down the fast path of `is_contiguous`.

I don't have a ton of context around `python_dispatch`, so please let me know if this is the intended use. The other way I imagined doing this is adding another enum to `SizesStridesPolicy`, like `PythonCustomStrides`, and then instead of checking both `is_python_dispatch()` and `sizes_strides_policy_ >= SizesStridesPolicy::CustomStrides`, I would just check `sizes_strides_policy_ == SizesStridesPolicy::PythonCustomStrides`

Differential Revision: [D36527481](https://our.internmc.facebook.com/intern/diff/D36527481)

### CODE REVIEW

**Comment by ezyang** [comment]
Looking over the implementation it looks like my outline was incomplete (also the outline predates policy consolidation), so there are a few more things to hit.

* After policy consolidation, CustomStrides also causes strides() to be virtualized. Strides will then hit `strides_custom` which will raise an error, as this is the default implementation. So if we do this, we also have to make strides overrideable by Python in the same way.
* This leads to a second problem, which is that once you flip on CustomStrides, there is literally no way to get at the underlying strides field on the TensorImpl object. Ordinarily, as a TensorImpl subclass you would have access to the protected `strides_default` method, but this is not an option for Python dispatch. It would be good if `super().strides()` did the right thing but I'm not exactly sure how to go about doing this. Maybe this is not a big deal as the point is you're opting out of the built in data.
* It would be better not to put the `is_python_dispatch` test directly inline in the `is_contiguous` method, as it is called from a lot of places and by putting it in the inline test you're forcing a bunch of extra cold code at all of the call sites, bloating icache. What is less certain is how exactly we should move it out of line. The minimal change is to place it inside the default virtual method implementation itself; this is cheap and easy to do but if a TensorImpl subclass overrides that virtual method it would be easy to forget to dispatch to Python if necessary and then this codepath would go dark (depending on the use case, though, this might not actually be a problem) If we need to guarantee this path gets hit for a subclass, no matter the TensorImpl, we could revive the non-virtual method which checks if `python_dispatch` before deciding what to do.

This also leaves me a question which is whether or not this is the best implementation strategy. An alternative implementation strategy would have been to add `is_contiguous` to `native_functions.yaml`, and then use `manual_cpp_binding` to maintain this method in TensorBase

```
  bool is_contiguous(at::MemoryFormat memory_format=at::MemoryFormat::Contiguous) const {
    return impl_->is_contiguous(memory_format);
  }
```

but have a test for "is Python subclass" here. But there are benefits to the approach in this PR: it is faster for the common case (since we only do one branch, not two) and it avoids having to deal with TorchScript related fallout from them having defined `is_contiguous` directly in TS primitives. So let's leave it as is.

File: c10/core/TensorImpl.h [hunk 1]
```diff
@@ -637,6 +637,16 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     if (C10_UNLIKELY(
             sizes_strides_policy_ >=
             static_cast<uint8_t>(SizesStridesPolicy::CustomStrides))) {
+      if (is_python_dispatch()) {
```
Comment (ezyang) [inline c10/core/TensorImpl.h:640]: Prefer this to be out-of-line, either in the virtual method or a separate `is_contiguous_slow` out-of-line nonvirtual method

File: c10/core/TensorImpl.h [hunk 2]
```diff
@@ -2215,6 +2225,11 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 
   Storage storage_;
 
+ public:
+  void set_custom_strides_policy() {
+    set_sizes_strides_policy(SizesStridesPolicy::CustomStrides);
```
Comment (ezyang) [inline c10/core/TensorImpl.h:2230]: `set_sizes_strides_policy` is public, might as well just use it directly (or if it's not public, let's make it public)

File: torch/csrc/autograd/python_variable.cpp [hunk 3]
```diff
@@ -546,6 +548,10 @@ static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, P
   // ```
   data.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
   data.set_requires_grad(r.toBool(2));
+  if (r.toBool(3)) {
+    data.unsafeGetTensorImpl()->set_python_dispatch(true);
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:552]: It shouldn't be necessary to do this, this should be computed automatically

File: torch/csrc/autograd/python_variable.cpp [hunk 4]
```diff
@@ -525,9 +527,9 @@ static PyObject* THPVariable_as_subclass(PyObject* _self, PyObject* args, PyObje
 static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, PyObject* kwargs) {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "_make_subclass(PyObject* cls, Tensor data, bool require_grad=False)",
+    "_make_subclass(PyObject* cls, Tensor data, bool require_grad=False, bool use_python_dispatch=False)",
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:530]: This is a misnomer; what you're really doing is making it possible to intercept strides

File: torch/csrc/autograd/python_variable.cpp [hunk 5]
```diff
@@ -1961,4 +1972,41 @@ c10::intrusive_ptr<TensorImpl> concrete_detach_fn(const c10::impl::PyInterpreter
   return res_t.getIntrusivePtr();
 }
 
+bool concrete_is_contiguous_fn(const c10::impl::PyInterpreter*, const c10::TensorImpl* self) {
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1975]: Since it looks like we're about to get a bunch of these, it would be a big help to figure out how to reduce the code duplication here. The biggest change between these functions is packing the arguments into an IValues list and then unpacking the result, everything else is generic.

File: c10/core/TensorImpl.h (hunk 2 above)
Comment (george-qi) [reply c10/core/TensorImpl.h:2230]: Sounds good - it's protected right now, so I'll just make it public!

**Comment by george-qi** [comment]
Thanks for the review and discussion in the meeting, super helpful!

> * After policy consolidation, CustomStrides also causes strides() to be virtualized. Strides will then hit `strides_custom` which will raise an error, as this is the default implementation. So if we do this, we also have to make strides overrideable by Python in the same way.

As discussed in the meeting, I'm currently limiting this diff to `is_contiguous` and the next to `contiguous` and punting on the others for now. Let me know if you feel strongly about supporting any of the other operations

> * This leads to a second problem, which is that once you flip on CustomStrides, there is literally no way to get at the underlying strides field on the TensorImpl object. Ordinarily, as a TensorImpl subclass you would have access to the protected `strides_default` method, but this is not an option for Python dispatch. It would be good if `super().strides()` did the right thing but I'm not exactly sure how to go about doing this. Maybe this is not a big deal as the point is you're opting out of the built in data.

Good call, and as you point out, for us, not having the underlying strides field is indeed okay for us

> * It would be better not to put the `is_python_dispatch` test directly inline in the `is_contiguous` method, as it is called from a lot of places and by putting it in the inline test you're forcing a bunch of extra cold code at all of the call sites, bloating icache. What is less certain is how exactly we should move it out of line. The minimal change is to place it inside the default virtual method implementation itself; this is cheap and easy to do but if a TensorImpl subclass overrides that virtual method it would be easy to forget to dispatch to Python if necessary and then this codepath would go dark (depending on the use case, though, this might not actually be a problem) If we need to guarantee this path gets hit for a subclass, no matter the TensorImpl, we could revive the non-virtual method which checks if `python_dispatch` before deciding what to do.

Thanks for the explanation and the learnings :) After a quick discussion with Christian, I've changed it the minimal change -- that is, putting it in the virtual method and leaving this up to the devs of tensor subclasses to add the Python dispatch if they'd like.

Please let me know if there are any other outstanding things!

File: torch/csrc/autograd/python_variable.cpp (hunk 5 above)
Comment (george-qi) [reply torch/csrc/autograd/python_variable.cpp:1997]: For now, I've just added a helper function between the duplicated code for `concrete_detach_fn` and `concrete_is_contiguous_fn`

File: torch/csrc/autograd/python_variable.cpp (hunk 4 above)
Comment (george-qi) [reply torch/csrc/autograd/python_variable.cpp]: Thanks for the comment - to stay in line with the C++ naming, I've changed this to `use_custom_strides`. Happy to change to any names others might prefer!

File: torch/csrc/autograd/python_variable.cpp [hunk 6]
```diff
@@ -525,9 +527,9 @@ static PyObject* THPVariable_as_subclass(PyObject* _self, PyObject* args, PyObje
 static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, PyObject* kwargs) {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "_make_subclass(PyObject* cls, Tensor data, bool require_grad=False)",
+    "_make_subclass(PyObject* cls, Tensor data, bool require_grad=False, bool use_custom_strides=False)",
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:530]: This should be a kwarg only; also minor bikeshed I suggest the name `dispatch_strides` instead.

**Review by ezyang — APPROVED** [review]

**Review by cpuhrsch — APPROVED** [review]

**Comment by malfet** [comment]
@pytorchbot revert this as this breaks dependency in https://github.com/pytorch/torchdistx, see this excerpt from internal build failure:
```
stderr: pytorch/torchdistx/src/cc/torchdistx/fake.cc:97:34: error: no matching constructor for initialization of 'c10::impl::PyInterpreter'
  FakePyInterpreter() noexcept : impl_{getName, decref, detach, dispatch} {}
```

**Comment by malfet** [comment]
By the way, look like this should have been marked as BC breaking?
Also, this failure (at least on compilation level) could have been easily avoided by adding default value to new argument

**Comment by albanD** [comment]
> By the way, look like this should have been marked as BC breaking?

I don't think so, torchdistx is using private APIs. So this is not BC breaking.

**Comment by malfet** [comment]
> > By the way, look like this should have been marked as BC breaking?
> 
> I don't think so, torchdistx is using private APIs. So this is not BC breaking.

Hmm, are all APIs in C10 consider to be private?

**Comment by george-qi** [comment]
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36527481).

**Comment by albanD** [comment]
> Hmm, are all APIs in C10 consider to be private?

That's a good question. At least our FC/BC guarantees don't apply to them. Just like we're "breaking" aten APIs all the time without warnings.
